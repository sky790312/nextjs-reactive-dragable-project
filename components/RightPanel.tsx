import React from 'react'
import styled from 'styled-components'
import ColorPicker from './ColorPicker'
import { Dispatch } from 'redux'
import { blockActions } from '@/store/state.block'
import { useSelector, shallowEqual, useDispatch } from 'react-redux'
import { RootState } from '@/store/index'

const RightPanel: React.FC = () => {
  const dispatch: Dispatch = useDispatch()
  const { setCurrentBlock } = blockActions
  const currentBlock = useSelector(
    (state: RootState) => state.block.currentBlock,
    shallowEqual,
  )

  const handleNumberChange = (
    e: React.ChangeEvent<HTMLInputElement>,
    property: 'x' | 'y' | 'o',
  ) => {
    const value = Number(e.target.value)
    const updateBlock = {
      ...currentBlock,
      [property]: value,
    }
    dispatch(setCurrentBlock(updateBlock))
  }

  const handleRangeChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const value = Math.round(Number(e.target.value)) / 100
    const updateBlock = {
      ...currentBlock,
      o: value,
    }
    dispatch(setCurrentBlock(updateBlock))
  }

  const handleColorChange = (color: string) => {
    const updateBlock = {
      ...currentBlock,
      c: color,
    }
    dispatch(setCurrentBlock(updateBlock))
  }

  return (
    <RightPanelWrapper shouldShow={!!currentBlock}>
      <Label>
        X{' '}
        <input
          type='number'
          min={0}
          max={999}
          value={currentBlock?.x ?? 0}
          onChange={e => handleNumberChange(e, 'x')}
        />
      </Label>
      <Label>
        Y{' '}
        <input
          type='number'
          min={0}
          max={999}
          value={currentBlock?.y ?? 0}
          onChange={e => handleNumberChange(e, 'y')}
        />
      </Label>
      <Label>
        O{' '}
        <input
          type='number'
          min={0}
          max={100}
          value={currentBlock?.o ?? 1}
          onChange={e => handleNumberChange(e, 'o')}
        />
        <input
          type='range'
          min={0}
          max={100}
          value={currentBlock?.o ? currentBlock?.o * 100 : 100}
          onChange={handleRangeChange}
        />
      </Label>
      <Label>
        B{' '}
        <ColorPicker
          defaultColor={currentBlock?.c}
          onColorChange={color => handleColorChange(color)}
          shouldShowText={true}
        />
      </Label>
    </RightPanelWrapper>
  )
}

export default RightPanel

const RightPanelWrapper = styled.div<{ shouldShow: boolean }>`
  padding: 8px;
  visibility: ${props => (props.shouldShow ? 'visible' : 'hidden')};
`

const Label = styled.label`
  display: grid;
  grid-template-columns: 16px auto minmax(0, 1fr);
  grid-gap: 8px;
`
