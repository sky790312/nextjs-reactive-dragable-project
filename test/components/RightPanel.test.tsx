import React from 'react'
import { render, withProviders } from '@/test/testUtils'
import RightPanel from '@/components/RightPanel'

describe('RightPanel component', () => {
  it('matches snapshot', () => {
    const { asFragment } = render(withProviders(<RightPanel />), {})
    expect(asFragment()).toMatchSnapshot()
  })
})
