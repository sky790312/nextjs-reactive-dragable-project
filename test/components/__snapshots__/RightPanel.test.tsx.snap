// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RightPanel component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="RightPanel__RightPanelWrapper-sc-1amlud5-0 dVOEbY"
  >
    <label
      class="RightPanel__Label-sc-1amlud5-1 eIbqoz"
    >
      X 
      <input
        max="999"
        min="0"
        type="number"
        value="0"
      />
    </label>
    <label
      class="RightPanel__Label-sc-1amlud5-1 eIbqoz"
    >
      Y 
      <input
        max="999"
        min="0"
        type="number"
        value="0"
      />
    </label>
    <label
      class="RightPanel__Label-sc-1amlud5-1 eIbqoz"
    >
      O 
      <input
        max="100"
        min="0"
        type="number"
        value="1"
      />
      <input
        max="100"
        min="0"
        type="range"
        value="100"
      />
    </label>
    <label
      class="RightPanel__Label-sc-1amlud5-1 eIbqoz"
    >
      B 
      <div
        class="ColorPicker__ColorPickerWrapper-sc-1fchp4r-0 eSLzde"
        color="#008800"
      >
        <input
          class="ColorPicker__ColorInput-sc-1fchp4r-1 gufyRf"
          type="color"
          value="#008800"
        />
      </div>
      #008800
    </label>
  </div>
</DocumentFragment>
`;
